### ISSUE TITLE
Use DNN by default for convolutions

### ISSUE THREAD

**Issue by ebenolson** [Lasagne/Lasagne#439]
Since `theano.tensor.nnet.conv2d` doesn't support `same` or explicit padding, the default `Conv2DLayer` implementation of padding can be quite inefficient - even if cuDNN is installed and Theano is using it under the hood. 

I think we should substitute `Conv2DDNNLayer` for `Conv2DLayer` automatically whenever possible.

**Comment by benanne** [comment]
That's difficult though, a lot of people probably don't have cuDNN installed. We were waiting for this, so we could just get rid of all our custom implementations (except maybe the cuda-convnet one): https://github.com/Theano/Theano/issues/2505

Unfortunately that seems to be taking quite a while.

Perhaps we could raise a warning when people try to use `Conv2DLayer` + 'same', and recommend another option instead? `Conv2DMMLayer` also supports it properly, and this is accessible to people who don't have cuDNN installed.

**Comment by Botev** [comment]
Would it be possible at least to include the dnn (and the cuda_convnet) module in the layers module in the official branch, so this can be done manually without having to edit the source code.

**Comment by benanne** [comment]
It already is? You don't need to edit any source code to use them.

**Comment by Botev** [comment]
What I meant is to include it in `lasagne/layers/__init__.py`. I can't reference it in any source code unless I know the absolute path to the module (or maybe there is some other magic trick to do it, which I'll be happy to hear). I'm not too expert on this kinda of Python topics, but I could not manage including it any other way

**Comment by benanne** [comment]
They used to be automatically imported, but this leads to problems when cudnn or cuda-convnet are not installed. We do not want users to be confronted with errors in this case -- only if they specifically import the modules that depend on these libraries. Surely it's not too much work to add an extra import if you specifically want to use these layers :)

**Comment by ebenolson** [comment]
It seems to work fine for Keras using `theano.sandbox.cuda.dnn.dnn_available`: https://github.com/fchollet/keras/pull/524/files

I'll do some profiling when I have a chance, but I think the difference can be pretty substantial.

**Comment by benanne** [comment]
Are we sure it does? Do they have tests for that part of the code?

I believe one of the motivations for not doing this was that accessing `theano.sandbox.cuda.dnn.dnn_available` fails when there is no CUDA GPU in the machine. Not entirely sure if I remember this correctly though.

**Comment by Botev** [comment]
You need to check before that theano.config.device.startswith("gpu"), which is the only case that the theano module imports the cuda drivers.

**Comment by nouiz** [comment]
That isn't exacts. Users could enable the GPU after import. This is needed
in some cases.

You need to import theano.sandbox.cuda before the check. This is probably
some people's think they must check the value of device as it so I'll do
the import automatically in that case.

If you have an idea of what to do to make this more clear it would be
great. If you didn't got this easily, then we must be missing something
about this. I'll check the doc of cuda_available. But if you think of other
places to clarify this, tell me.
Le 28 sept. 2015 06:50, "Alexander Botev" <notifications@github.com> a
écrit :

**Comment by benanne** [comment]
But we recently established that this check is too strict: https://github.com/Lasagne/Lasagne/issues/379

**Comment by f0k** [comment]
> We were waiting for this, so we could just get rid of all our custom implementations (except maybe the cuda-convnet one): Theano/Theano#2505

The implementation of that is in: https://github.com/Theano/Theano/pull/2665
I was thinking of doing it myself in spring, but then I learned somebody is already working on this and will probably be finished within a few weeks (that was in March). I'm still thinking about just doing it myself to finally get rid of all the problems, but it was easy enough to just switch to `Conv2DDNNLayer`.

> That isn't exacts. Users could enable the GPU after import.

So we can't make the auto-import dependent on whether the GPU is used, we have to make it dependent on whether it *could* be used. That would probably be `theano.sandbox.cuda.cuda_available and theano.sandbox.cuda.dnn.dnn_available()`.

> I can't reference it in any source code unless I know the absolute path to the module (or maybe there is some other magic trick to do it, which I'll be happy to hear).

Is the magic trick you're looking for `from lasagne.layers.dnn import Conv2DDNNLayer`?

**Comment by nouiz** [comment]
All tests for 2665 pass. It have been hard to do as there was many corner
case to fix. It was heavily tested, that explain why it was much longer.
One example of problem found is that the grad test with cudnn do not always
pass. This was due that some of the implementation aren't determinitic.
When we get not deterministic errors, it get much harder to find!

Anyway, I was told it is mostly code clean up left. I won't give a timeline
this time, just in case I get it wrong again, but I'm optimistic :)

**Comment by benanne** [comment]
Good to know, thanks! :)

**Comment by davidbau** [comment]
Just chiming in here - new to Lasagne after using Cafe, Keras, etc, and it took several day for me to figure out why Lasagne was the only framework that didn't use my GPU by default (so slow!)  Turns out that I had to switch to use Conv2DDNNLayer, but discovering that was the problem was difficult to discover.

**Comment by benanne** [comment]
That doesn't make sense, Lasagne's regular `Conv2DLayer` should also use the GPU by default, and it should use cuDNN when it is installed. If it doesn't/didn't, something might be wrong with your installation.

**Comment by davidbau** [comment]
I spent a couple days reinstalling the whole stack several times, and didn't have trouble with other frameworks.

Is it possibly due to my using pad="same"?

Although I think I saw the problem even when reducing it to one layer without a padding option.  Happy to spend a little time trying to diagnose if it's possible it's a Lasagne bug.

**Comment by benanne** [comment]
`pad='same'` will be slower with `Conv2DLayer` than with `Conv2DDNNLayer` because the padding is explicit in the former case (i.e. memory is allocated for it), and not in the latter. But it should definitely still run on the GPU! How did you establish that it did not? Did you use profile mode?

**Comment by davidbau** [comment]
Interesting.  No, it definitely did not use the GPU when using Conv2DLayer.

My GPU was at 0% during several minutes of training on a series of (small 256 image) batches, while monitoring with nvidia-smi.  When it uses the GPU, it's very visible with the same tool.  (And each batch also only takes a few seconds.)

**Comment by benanne** [comment]
That is very odd indeed. It would be very helpful to be able to reproduce this. Do you have a script that we could run?

**Comment by davidbau** [comment]
Here you go: I've tried to make a branch that rolls back to where we were having the problem, and make the setup self-contained so you can repro.  Note it's on branch **gpubug**

```
git clone https://github.com/jonhoo/periscope.git
cd periscope
git checkout gpubug
make solve
```

**Comment by benanne** [comment]
I seem to be unable to get this to work because of missing dependencies. It was complaining about termcolor, which I've installed with `env/bin/pip install termcolor`, but then it complained about `progressbar` as well, which I am unable to install in this way.

All-in-all I was hoping for something a little more self-contained :) This makefile had me downloading a 450MB data file. But of course I understand if you don't have time for anything like that right now.

**Comment by ebenolson** [comment]
I tried it and it seems to be using the GPU to me.

**Comment by f0k** [comment]
If you tried Keras and it was using the GPU, there must be something you did differently -- both Lasagne and Keras build on top of Theano. Did you specify `THEANO_FLAGS=device=gpu,floatX=float32` or the `.theanorc` equivalent? Are you sure your network input is single precision (float32)?

**Comment by davidbau** [comment]
Thanks for looking and sorry about the biggish data download; that just happens to be what I'm working with.  termcolor and other dependencies are supposed to be brought in via the "env.sh" rule in the Makefile - running everything in a virtualenv with those pieces.   I do have a .theanorc with exactly those settings, and the data is float32.  Remember, everything works when I set things to the Conv2DDNN versions.

It's not urgent for me since I have a perfectly workable workaround, just mysterious, but if I get some more time I will try to narrow it down further.

**Comment by benanne** [comment]
I did run env.sh but it looks like it didn't install those things. I could install `termcolor` manually, but if I try to install `progressbar` pip complains. I think this is a known Python 3 compatibility issue.

To be 100% certain that it's not running on the GPU as you say, could you provide the output of running the code with `THEANO_FLAGS=profile=True`?

**Comment by ebenolson** [comment]
Definitely something odd going on...
Reduced example: https://gist.github.com/cc6b319c99ff3f7d4d94

Seems to be related to the gradient computation, and only triggered with `stride != 1`

**Comment by benanne** [comment]
Yeah, the problem is the stride. It triggers Theano to use `theano.tensor.nnet.ConvGrad3D.ConvGrad3D` for the gradient, which is a CPU implementation (and a slow one to boot). For some reason I thought this was fixed, but I guess it isn't. With strided convolutions you definitely want to use Conv2DMM or Conv2DDNN for now. Hopefully the new `abstract_conv2d` interface also fixes this issue.

**Comment by davidbau** [comment]
Thanks for the tips on stride!  And thanks for wading through my example.

**Comment by f0k** [comment]
> Hopefully the new `abstract_conv2d` interface also fixes this issue.

Yes, using the placeholder Ops does allow the convolution *and its gradients* to be automatically replaced with the cuDNN-based one. Maybe we should help the Theano devs making the switch; I'll have a look at what's missing after finishing a review (of a journal article this time, not a PR ;) ).

**Comment by ebenolson** [comment]
I tried `abstract_conv2d` the other day and while it's definitely an improvement, it [still doesn't seem to use cuDNN](https://gist.github.com/anonymous/b2285c6916865bd2e80c). Is that just not built yet or do I need to do something to enable the optimization?

**Comment by f0k** [comment]
> Is that just not built yet

Funny, I would have expected this to work. There is a suitable optimizer in Theano: https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/dnn.py#L2444
Maybe they're not registered in the correct order, so the GpuCorrMM kicks in first?
@nouiz, any idea? For Eben, the cuDNN convolution doesn't take precedence over the gemm-based convolution in his gist: https://gist.github.com/anonymous/b2285c6916865bd2e80c

**Comment by nouiz** [comment]
I have a fix here:

https://github.com/Theano/Theano/pull/3667

Just a bad condition...

**Comment by f0k** [comment]
Closing this, both Theano and Lasagne have been updated in the meantime such that Conv2DLayer will always use cuDNN if installed and not explicitly disabled.
